x/slashing/keeper/signing_info.go

```diff
@@ -141,20 +95,133 @@ func (k Keeper) IsTombstoned(ctx sdk.Context, consAddr sdk.ConsAddress) bool {
 	return signInfo.Tombstoned
 }
 
-// SetValidatorMissedBlockBitArray sets the bit that checks if the validator has
-// missed a block in the current window
-func (k Keeper) SetValidatorMissedBlockBitArray(ctx sdk.Context, address sdk.ConsAddress, index int64, missed bool) {
+// GetMissedBlockBitmapChunk gets the bitmap chunk at the given chunk index for
+// a validator's missed block signing window.
+func (k Keeper) GetMissedBlockBitmapChunk(ctx sdk.Context, addr sdk.ConsAddress, chunkIndex int64) []byte {
 	store := ctx.KVStore(k.storeKey)
-	bz := k.cdc.MustMarshal(&gogotypes.BoolValue{Value: missed})
-	store.Set(types.ValidatorMissedBlockBitArrayKey(address, index), bz)
+	chunk := store.Get(types.ValidatorMissedBlockBitmapKey(addr, chunkIndex))
+	return chunk
 }
 
-// clearValidatorMissedBlockBitArray deletes every instance of ValidatorMissedBlockBitArray in the store
-func (k Keeper) clearValidatorMissedBlockBitArray(ctx sdk.Context, address sdk.ConsAddress) {
+// SetMissedBlockBitmapChunk sets the bitmap chunk at the given chunk index for
+// a validator's missed block signing window.
+func (k Keeper) SetMissedBlockBitmapChunk(ctx sdk.Context, addr sdk.ConsAddress, chunkIndex int64, chunk []byte) {
```

Comment:
It looks like these two methods should be unexported. They are only called from within this package. The chunking looks like an implementation detail that should not be part of the keeper's exported API. External callers should be using `{Get,Set}MissedBlockBitmapValue`.

I also wonder whether these two methods should operate on `[types.MissedBlockBitmapChunkSize]byte` instead of a plain slice, for a little more compile-time safety. I'm not sure what would go wrong if an incorrect-sized slice was passed.

Reply:
Yeah there's no need for them to be public. Updated.

Re; array vs slice, the size of the bytes is not actually `MissedBlockBitmapChunkSize`, it could be a bit more...depending on the value of the chunk size, the bitmap might need to be padded slightly to accommodate the correct number of int64 words.

If we stick with the current value ``MissedBlockBitmapChunkSize`, I'm pretty sure the bytes will always be the same fixed size, but if we ever change `MissedBlockBitmapChunkSize` then we'd have to change the array values.

Do you really see a strong motivation for this?

Reply:
It seems feasible that the constant chunk size could change.

If there is a programmer mistake where they end up passing an incorrect byte slice, what happens? Would it make sense to add a runtime check that the slice has the correct length?